File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.*;

class Solution {
    public int solution(String[][] book_time) {
        Arrays.sort(book_time, (a, b) -> {
            if (a[0].equals(b[0])) return a[1].compareTo(b[1]);
            return a[0].compareTo(b[0]);
        });

        PriorityQueue<Integer> rooms = new PriorityQueue<>();

        for (String[] time : book_time) {
            int start = toMinute(time[0]);
            int end = toMinute(time[1]) + 10;

            if (!rooms.isEmpty() && rooms.peek() <= start) {
                rooms.poll();
            }

            rooms.offer(end);
        }

        return rooms.size();
    }

    private int toMinute(String time) {
        String[] t = time.split(":");
        return Integer.parseInt(t[0]) * 60 + Integer.parseInt(t[1]);
    }
}

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.js
function solution(book_time) {
    book_time.sort((a, b) => {
        if (a[0] === b[0]) return a[1].localeCompare(b[1]);
        return a[0].localeCompare(b[0]);
    });
    
    function toMinute(time){
        const [h, m] = time.split(":").map(Number);
        return h * 60 + m;
    }
    
    var rooms = [];
    for (const [startStr, endStr] of book_time) {
        const start = toMinute(startStr);
        const end = toMinute(endStr) + 10;
        
        let flag = false;
        for(let i = 0; i < rooms.length; i++){
            if(rooms[i] <= start){
                flag = true;
                rooms[i] = end;
                break;
            }
        }
        if(!flag){
            rooms.push(end);
        }
    }
    
    return rooms.length;
}